sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionAllColumnFamiliesReaderSuite.scala

```diff
@@ -533,4 +643,399 @@ class StatePartitionAllColumnFamiliesReaderSuite extends StateDataSourceTestBase
+
+        Seq("right-keyToNumValues", "left-keyToNumValues").foreach(colFamilyName => {
+          val normalDf = getNormalReadDf(tempDir.getAbsolutePath, Option(colFamilyName))
+
+          val keyToNumValuesKeySchema = StructType(Array(
+            StructField("key", IntegerType)
+          ))
+          val keyToNumValueValueSchema = StructType(Array(
+            StructField("value", LongType)
+          ))
+
+          compareNormalAndBytesData(
+            normalDf.collect(),
+            stateBytesDf.collect(),
+            colFamilyName,
+            keyToNumValuesKeySchema,
+            keyToNumValueValueSchema)
+        })
+
+        Seq("right-keyWithIndexToValue", "left-keyWithIndexToValue").foreach(colFamilyName => {
+          val normalDf = getNormalReadDf(tempDir.getAbsolutePath, Option(colFamilyName))
+          val keyToNumValuesKeySchema = StructType(Array(
+            StructField("key", IntegerType, nullable = false),
+            StructField("index", LongType)
+          ))
+          val keyToNumValueValueSchema = StructType(Array(
+              StructField("value", IntegerType, nullable = false),
+              StructField("time", TimestampType, nullable = false),
+              StructField("matched", BooleanType)
+          ))
+
+          compareNormalAndBytesData(
+            normalDf.collect(),
+            stateBytesDf.collect(),
+            colFamilyName,
+            keyToNumValuesKeySchema,
+            keyToNumValueValueSchema)
+        })
+      }
+    }
+  }
+
+  test("SPARK-54419: stream-stream joinV3") {
+    testStreamStreamJoinV3()
+  }
+}
+
+/**
+ * Stateful processor with multiple state variables (ValueState + ListState)
+ * for testing multi-column family reading.
+ */
+class MultiStateVarProcessor extends StatefulProcessor[String, String, (String, String)] {
+  @transient private var _countState: ValueState[Long] = _
+  @transient private var _itemsList: ListState[String] = _
+  @transient private var _itemsMap: MapState[String, SimpleMapValue] = _
+
+  override def init(outputMode: OutputMode, timeMode: TimeMode): Unit = {
+    _countState = getHandle.getValueState[Long]("countState", Encoders.scalaLong, TTLConfig.NONE)
+    _itemsList = getHandle.getListState[String]("itemsList", Encoders.STRING, TTLConfig.NONE)
+    _itemsMap = getHandle.getMapState[String, SimpleMapValue](
+      "itemsMap", Encoders.STRING, Encoders.product[SimpleMapValue], TTLConfig.NONE)
+  }
+
+  override def handleInputRows(
+          key: String,
+          inputRows: Iterator[String],
+          timerValues: TimerValues): Iterator[(String, String)] = {
+    val currentCount = Option(_countState.get()).getOrElse(0L)
+    var newCount = currentCount
+    inputRows.foreach { item =>
+      newCount += 1
+      _itemsList.appendValue(item)
+      _itemsMap.updateValue(item, SimpleMapValue(newCount.toInt))
+    }
+    _countState.update(newCount)
+    Iterator((key, newCount.toString))
+  }
+}
+
```

Comment:
Is there a reason why you're implementing your own processors instead of reusing the ones in TransformWithStateSuite: https://github.com/apache/spark/blob/25c57d87382ca90850541d4d20447bfd842ec85a/sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala#L606